# Advanced Data Manipulation     &#x1f4c8; <br>

Hello, and welcome to my independent lab for week 8. In this lab we will go over some advanced data manipulation! <br>
We will be working with the 'gh_turnover_four_projects.csv' file and do some renaming of columns and some data wrangling     &#x1f920; <br>

## Setup

In [2]:
 # Import Modules
    
import os
import numpy as np
import pandas as pd

# Check the current working directory

os.getcwd()

'/home/school/Documents/python/week_8'

Awesome, we are in the correct current working directory! Now let's get the file read into python: 

In [5]:

#Reads in the provided .csv file

gh_turnover = pd.read_csv("data/gh_turnover_four_projects.csv")

#Print to confirm

gh_turnover.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 473 entries, 0 to 472
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   project_id     473 non-null    int64  
 1   domain         473 non-null    object 
 2   language       473 non-null    object 
 3   windows        473 non-null    int64  
 4   window_idx     473 non-null    int64  
 5   num_team       473 non-null    int64  
 6   num_commits    473 non-null    int64  
 7   blau_gender    473 non-null    float64
 8   Gini_gh_ten    473 non-null    float64
 9   user_id        473 non-null    int64  
 10  Core1          473 non-null    int64  
 11  commits        473 non-null    int64  
 12  propCommits    473 non-null    float64
 13  gender         449 non-null    object 
 14  github_tenure  473 non-null    int64  
 15  leavesNextQ    473 non-null    int64  
dtypes: float64(3), int64(10), object(3)
memory usage: 59.2+ KB


And just like that, we have the data in python!

## Renaming <br>
In this segment, we will rename the columns in the following manner:  

'domain' -> 'project_domain' <br>
'language' -> 'project_language' <br>
'windows' -> 'project_age' <br>
'window_idx' -> 'quarter' <br>
'num_team' -> 'team_size' <br>
'num_commits' -> 'project_commits' <br>
'blau_gender' -> 'gender_blau' <br>
'Gini_gh_ten' -> 'gh_tenure_gini' <br>
'Core1' -> 'core_dev' <br>
'commits' -> 'user_commits' <br>
'propCommits' -> 'commit_proportion' <br>
'github_tenure' -> 'gh_tenure' <br>
'leavesNextQ' -> 'leaves_next_q'

In [10]:

# Renames each columns to the assigned name

gh_turnover = (gh_turnover
.rename(columns={
    'domain':'project_domain',
    'language' : 'project_language',
    'windows' : 'project_age',
    'window_idx' : 'quarter',
    'num_team' : 'team_size',
    'num_commits' : 'project_commits',
    'blau_gender' : 'gender_blau',
    'Gini_gh_ten' : 'gh_tenure_gini',
    'Core1' : 'core_dev',
    'commits' : 'user_commits',
    'propCommits' : 'commit_proportion',
    'github_tenure' : 'gh_tenure',
    'leavesNextQ' : 'leaves_next_q'
})
)


## Data Wrangling     &#x1f4bb; <br>
Now that we have gone through and renamed the columns in the dataset, let's now get to some data manipulation!

In [13]:

#Calculates the mean and median of 'user_commits'

(gh_turnover
.agg({
    'user_commits':['mean','median']
}))


user_commits
mean      339.365751
median     18.000000

In [41]:

#Selects all the columns that start with "p" or contain the letter "g". Assigns it to a new dataframe.

turnover_new = (gh_turnover
               .filter(regex = '(^p|g)'))

#Print to confirm

turnover_new.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 473 entries, 0 to 472
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   project_id        473 non-null    int64  
 1   project_domain    473 non-null    object 
 2   project_language  473 non-null    object 
 3   project_age       473 non-null    int64  
 4   project_commits   473 non-null    int64  
 5   gender_blau       473 non-null    float64
 6   gh_tenure_gini    473 non-null    float64
 7   gender            449 non-null    object 
 8   gh_tenure         473 non-null    int64  
dtypes: float64(2), int64(4), object(3)
memory usage: 33.4+ KB


In [49]:

#Selects the columns in which 'project_commits' is greater than 9,000

(turnover_new
.query("project_commits > 9000"))



project_id project_domain project_language  project_age  project_commits  \
0          1101          OTHER           Python           23            22688   
1          1101          OTHER           Python           23            22688   
2          1101          OTHER           Python           23            22688   
3          1101          OTHER           Python           23            22688   
4          1101          OTHER           Python           23            22688   
..          ...            ...              ...          ...              ...   
116        1101          OTHER           Python           23            14210   
117        1101          OTHER           Python           23            14210   
118        1101          OTHER           Python           23            14210   
119        1101          OTHER           Python           23            14210   
120        1101          OTHER           Python           23            14210   

     gender_blau  gh_tenure_gini gender  gh_tenure  
0            0.0        0.422259   male        426  
1            0.0        0.422259   male         46  
2            0.0        0.422259   male        941  
3            0.0        0.422259   male       1870  
4            0.0        0.422259   male        781  
..           ...             ...    ...        ...  
116          0.0        0.411372   male        495  
117          0.0        0.411372   male        487  
118          0.0        0.411372   male        317  
119          0.0        0.411372   male        418  
120          0.0        0.411372   male        179  

[114 rows x 9 columns]

In [48]:

#Performs the previous two actions in one swoop:

    #Selects columns that either start with 'p' or contain 'g'
    #Selects rows in which 'project_commits > 9000'
    
turnover_newer= (gh_turnover
.filter(regex='(^p|g)')
.query('project_commits > 9000'))
    
#Print to confirm
turnover_newer
    

project_id project_domain project_language  project_age  project_commits  \
0          1101          OTHER           Python           23            22688   
1          1101          OTHER           Python           23            22688   
2          1101          OTHER           Python           23            22688   
3          1101          OTHER           Python           23            22688   
4          1101          OTHER           Python           23            22688   
..          ...            ...              ...          ...              ...   
116        1101          OTHER           Python           23            14210   
117        1101          OTHER           Python           23            14210   
118        1101          OTHER           Python           23            14210   
119        1101          OTHER           Python           23            14210   
120        1101          OTHER           Python           23            14210   

     gender_blau  gh_tenure_gini gender  gh_tenure  
0            0.0        0.422259   male        426  
1            0.0        0.422259   male         46  
2            0.0        0.422259   male        941  
3            0.0        0.422259   male       1870  
4            0.0        0.422259   male        781  
..           ...             ...    ...        ...  
116          0.0        0.411372   male        495  
117          0.0        0.411372   male        487  
118          0.0        0.411372   male        317  
119          0.0        0.411372   male        418  
120          0.0        0.411372   male        179  

[114 rows x 9 columns]

### Original Data Wrangling <br>
In this segment, we will be working with the original "gh_turnover" dataframe, which was created from the provided file.

In [51]:

#Selects columns that end with 'e' or contains the letter 'q', and have 'gh_tenure' greater than or equal to 1991.

(gh_turnover
.filter(regex='(q|e$)')
.query("gh_tenure > 1991"))


project_language  project_age  quarter  team_size  gh_tenure  \
339             Java           16       17          2       2006   
457           Python           23       24         20       2054   

     leaves_next_q  
339              0  
457              0

In [52]:

#Calculates the mean and the median of 'user_commits' 

(gh_turnover
.query("gh_tenure > 1991")
.agg({
    'user_commits':['mean','median']
}))


user_commits
mean             3.5
median           3.5

#### Comparison <br>

We can see that the mean and median of the 'user_commits" for the unqueried dataset is ~339.37 and 18 respectively. Now with the queried dataset both the mean and the median are 3.5. These aggregate values are considerably lower, and this makes sense, as the law of small numbers applies. Let's take a look at the number of observations for the original and queried dataset:

In [60]:

# Calculates the number of rows for the original dataset

gh_turnover.shape[0]


473

In [65]:
# Calculates the number of rows for the queried dataset

(gh_turnover
.query("gh_tenure > 1991")
.agg({'project_id':['count']}))

project_id
count           2

We can see that there are only two rows in the queried dataset, thus supporting the argument that this is due to the law of small numbers. There are very few observations in this dataset, so the mean and median being the same small value makes sense. 